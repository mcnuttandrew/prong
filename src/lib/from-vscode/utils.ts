import * as Json from "jsonc-parser";
import { JSONSchemaRef, JSONSchema } from "../JSONSchemaTypes";
import { ASTNode } from "./parser";
import * as nls from "vscode-nls";
let localize = nls.loadMessageBundle();

export function isNumber(val: any): val is number {
  return typeof val === "number";
}

export function isDefined(val: any): val is object {
  return typeof val !== "undefined";
}

export function isBoolean(val: any): val is boolean {
  return typeof val === "boolean";
}

export function isString(val: any): val is string {
  return typeof val === "string";
}

export function getNodeValue(node: ASTNode): any {
  return Json.getNodeValue(node);
}

export function asSchema(
  schema: JSONSchemaRef | undefined
): JSONSchema | undefined {
  if (isBoolean(schema)) {
    return schema ? {} : { not: {} };
  }
  return schema;
}

export function contains(
  node: ASTNode,
  offset: number,
  includeRightBound = false
): boolean {
  return (
    (offset >= node.offset && offset < node.offset + node.length) ||
    (includeRightBound && offset === node.offset + node.length)
  );
}

export function equals(one: any, other: any): boolean {
  if (one === other) {
    return true;
  }
  if (
    one === null ||
    one === undefined ||
    other === null ||
    other === undefined
  ) {
    return false;
  }
  if (typeof one !== typeof other) {
    return false;
  }
  if (typeof one !== "object") {
    return false;
  }
  if (Array.isArray(one) !== Array.isArray(other)) {
    return false;
  }

  var i: number, key: string;

  if (Array.isArray(one)) {
    if (one.length !== other.length) {
      return false;
    }
    for (i = 0; i < one.length; i++) {
      if (!equals(one[i], other[i])) {
        return false;
      }
    }
  } else {
    var oneKeys: string[] = [];

    for (key in one) {
      oneKeys.push(key);
    }
    oneKeys.sort();
    var otherKeys: string[] = [];
    for (key in other) {
      otherKeys.push(key);
    }
    otherKeys.sort();
    if (!equals(oneKeys, otherKeys)) {
      return false;
    }
    for (i = 0; i < oneKeys.length; i++) {
      if (!equals(one[oneKeys[i]], other[oneKeys[i]])) {
        return false;
      }
    }
  }
  return true;
}

export enum ErrorCode {
  Undefined = 0,
  EnumValueMismatch = 1,
  Deprecated = 2,
  UnexpectedEndOfComment = 0x101,
  UnexpectedEndOfString = 0x102,
  UnexpectedEndOfNumber = 0x103,
  InvalidUnicode = 0x104,
  InvalidEscapeCharacter = 0x105,
  InvalidCharacter = 0x106,
  PropertyExpected = 0x201,
  CommaExpected = 0x202,
  ColonExpected = 0x203,
  ValueExpected = 0x204,
  CommaOrCloseBacketExpected = 0x205,
  CommaOrCloseBraceExpected = 0x206,
  TrailingComma = 0x207,
  DuplicateKey = 0x208,
  CommentNotPermitted = 0x209,
  SchemaResolveError = 0x300,
}
