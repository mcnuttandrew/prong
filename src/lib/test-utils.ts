import { SyntaxNode } from "@lezer/common";
import { parser } from "@lezer/json";
// could also write one that is find by text content?
export function findNodeByLocation(
  text: string,
  from: number,
  to: number
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      if (node.from === from && node.to === to) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}

export function findNodeByText(
  text: string,
  matchText: string
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      const testText = text.slice(node.from, node.to);
      if (testText === matchText) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}
