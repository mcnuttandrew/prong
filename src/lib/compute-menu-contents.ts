import isequal from "lodash.isequal";
import { SyntaxNode } from "@lezer/common";

import { MenuEvent, boundCheck } from "./modify-json";
import { SchemaMap } from "../components/Editor";
import { JSONSchema7, JSONSchema7Definition } from "json-schema";
import { simpleParse } from "./utils";

// type JSONSchema = any;

export type MenuRow = { label: string; elements: MenuElement[] };
export type MenuElement =
  | {
      type: "button";
      label?: string;
      content: string;
      onSelect: MenuEvent;
    }
  | {
      type: "dropdown";
      content: string[];
      onSelect: MenuEvent;
    }
  | { type: "display"; label?: string; content: string }
  | { type: "free-input"; label: string }
  | { type: "projection"; label?: string; element: JSX.Element };
interface ComponentProps {
  content: JSONSchema7;
  node: SyntaxNode;
  fullCode: string;
}
type Component = (props: ComponentProps) => MenuRow[];
type componentContainer = Record<string, Component>;

export const nodeToId = (node: SyntaxNode): `${number}-${number}` =>
  `${node.from}-${node.to}`;
const EnumPicker: Component = (props) => {
  const { content, node } = props;
  // TODO dont switch if this is the current value?
  return [
    {
      label: "content",
      elements: (content.enum as any[]).map((val: string) => ({
        type: "button",
        content: val,
        onSelect: {
          type: "simpleSwap",
          payload: `"${val}"`,
          nodeId: nodeToId(node),
        },
      })),
    },
  ];
};

function simpleFillOut(content: JSONSchema7) {
  const simpleTypes: Record<string, string> = {
    string: "",
    object: `{}`,
    number: "0",
    boolean: "true",
  };
  if ((content as any).type in simpleTypes) {
    return simpleTypes[(content as any).type];
  } else if (content.anyOf && content.anyOf.length) {
    const childTypes = content.anyOf.map((x: any) => x.type);
    return childTypes.every((x: string) => childTypes[0] === x) &&
      childTypes[0] in simpleTypes
      ? simpleTypes[childTypes[0]]
      : null;
  } else if (content.enum) {
    // doesn't do anything rn
    return content.enum.filter((x) => typeof x === "string")[0];
  } else {
    return null;
  }
}

const parseContent = (node: SyntaxNode, fullCode: string, defaultVal?: any) => {
  const slice = fullCode.slice(node.from, node.to);
  return simpleParse(slice, defaultVal);
};

const ObjPicker: Component = (props) => {
  const { content, node, fullCode } = props;
  const containerNode = getContainingObject(node);
  const parsedContent = parseContent(containerNode, fullCode, {});
  const currentKeys = new Set(Object.keys(parsedContent));
  // TODO this gets this wrong if coming from { / }
  const addFieldEntries: MenuElement[] = Object.entries(
    content.properties || {}
  )
    .filter((x) => !currentKeys.has(x[0]))
    .map(([content, prop]: any) => {
      return {
        type: "button",
        content,
        onSelect: {
          type: "addObjectKey",
          payload: {
            key: `"${content}"`,
            value: `${simpleFillOut(prop)}` as any,
          },
          nodeId: nodeToId(node),
        },
      };
    });
  return [
    addFieldEntries.length && {
      label: "Add",
      elements: addFieldEntries,
    },
    {
      label: "Add Field",
      elements: [
        {
          type: "free-input",
          label: "Add field",
          onSelect: {
            type: "addObjectKey",
            nodeId: nodeToId(node),
            payload: { key: "$$INPUT_BIND", value: "null" },
          },
        },
      ],
    },
  ].filter((x) => x) as MenuRow[];
};

// TODO flatten nested anyOfs and remove duplicates
function flattenAnyOf(content: JSONSchema7): any {
  if (!content || !content.anyOf) {
    return content;
  }
  return content.anyOf.reduce((acc: any[], row: any) => {
    if (row.description) {
      acc.push({ description: row.description });
    }
    return acc.concat(row.anyOf ? flattenAnyOf(row) : row);
  }, []);
}

function removeDupsInAnyOf(content: JSONSchema7[]) {
  return content.filter((row, idx) =>
    content.slice(idx + 1).every((innerRow) => !isequal(row, innerRow))
  );
}

function bundleConstsToEnum(content: JSONSchema7[]) {
  const consts = content.filter((x) => x.const);
  const nonConsts = content.filter((x) => !x.const);
  return [...nonConsts, { enum: consts.map((x) => x.const) }];
}

// goes anything not an object up to it's container, only to be used below
const getContainingObject = (node: SyntaxNode): SyntaxNode => {
  return node.type.name === "Object" ? node : getContainingObject(node.parent!);
};

const getUsedPropertiesForContainer = (node: SyntaxNode): SyntaxNode[] => {
  const props: SyntaxNode[] = [];
  let pointer = node.firstChild;
  while (pointer) {
    props.push(pointer);
    pointer = pointer.nextSibling;
  }
  return props;
};

function materializeRequiredProps(content: any): string[] {
  const requiredProps = new Set<string>(
    Array.isArray(content.required) ? content.required : []
  );
  return Array.from(requiredProps);
}
function materializeAnyOfOption(content: any): string {
  const requiredPropsArr = materializeRequiredProps(content);
  return JSON.stringify(
    Object.fromEntries(
      requiredPropsArr.map((x) => [
        x,
        x in (content?.properties || {})
          ? simpleFillOut(content.properties[x])
          : "null",
      ])
    )
  );
}

function AnyOfObjOptionalFieldPicker(
  content: any,
  node: SyntaxNode,
  fullCode: string
): MenuRow[] {
  const requiredPropsArr = materializeRequiredProps(content);
  const requiredPropObject = materializeAnyOfOption(content);

  const isObject: boolean = new Set(["{", "}", "Object"]).has(node.type.name);
  const addProps = deduplicateAndSortArray(
    Object.keys(content?.properties || {}).concat(requiredPropsArr)
  );

  const containerNode = getContainingObject(node);
  const containerProps = getUsedPropertiesForContainer(containerNode);
  const slice = fullCode.slice(containerNode.from, containerNode.to);
  const inUseKeys = new Set(Object.keys(simpleParse(slice, {})));
  return [
    content.$$labeledType && {
      label: "JSON Schema Type",
      elements: [{ type: "display", content: content.$$labeledType }],
    },
    !isObject && {
      label: "Switch to",
      elements: [
        requiredPropsArr.length > 0 && {
          type: "button",
          onSelect: {
            type: "simpleSwap",
            nodeId: nodeToId(node),
            payload: requiredPropObject,
          },
          label: requiredPropObject,
          content: `object`,
        },
        requiredPropsArr.length === 0 && {
          type: "button",
          onSelect: {
            type: "simpleSwap",
            nodeId: nodeToId(node),
            payload: "{}",
          },
          content: "empty object",
        },
      ],
    },
    isObject && {
      label: "Remove",
      elements: Array.from(inUseKeys).map((x, idx) => ({
        type: "button",
        content: x,
        onSelect: {
          type: "removeObjectKey",
          nodeId: nodeToId(containerProps[idx + 1]),
        },
      })),
    },
    isObject && {
      label: "Add",
      elements: addProps
        .filter((x) => !inUseKeys.has(x))
        .map((x) => ({
          type: "button",
          content: x,
          onSelect: {
            type: "addObjectKey",
            nodeId: nodeToId(node),
            payload: { key: `"${x}"`, value: "null" },
          },
        })),
    },
  ].filter((x) => x) as MenuRow[];
}

const deduplicateAndSortArray = (arr: string[]): string[] => {
  return Array.from(new Set(arr)).sort((a, b) => a.localeCompare(b));
};

function AnyOfArray(content: JSONSchema7, node: SyntaxNode): MenuRow[] {
  const numElements = 5;
  const arrayTypeDefaults: any = {
    boolean: true,
    string: "",
    number: 0,
    object: {},
    array: [],
  };
  const arrayType = (content?.items as any)?.type;

  if (arrayType) {
    return [
      {
        label: "Switch to",
        elements: [
          {
            type: "button",
            content: "Empty array",

            onSelect: {
              type: "simpleSwap",
              nodeId: nodeToId(node),
              payload: "[]",
            },
          },
        ],
      },
    ];
  }
  // i think this can be dropped in favor of just no return?
  const payload = JSON.stringify(
    [...new Array(numElements)].map(() => arrayTypeDefaults[arrayType])
  );
  return [
    {
      label: "arrrX",
      elements: [
        {
          type: "button",
          content: `Switch to array of ${numElements} ${JSON.stringify(
            arrayTypeDefaults[arrayType]
          )}s`,
          onSelect: { type: "simpleSwap", nodeId: nodeToId(node), payload },
        },
      ],
    },
  ];
}

const AnyOfPicker: Component = (props) => {
  const { content, node, fullCode } = props;
  const simpleType = new Set(["string", "number", "boolean", "null"]);
  const simpleTypeMap: Record<string, string> = {
    string: '""',
    number: "0",
    boolean: "true",
    null: "null",
  };
  const currentNodeType = typeof simpleParse(
    props.fullCode.slice(node.from, node.to),
    {}
  );
  const anyOptions = bundleConstsToEnum(
    removeDupsInAnyOf(flattenAnyOf(content))
  );
  const rows = anyOptions.flatMap((opt: any, idx) => {
    const isOfDescribedType =
      (opt?.type || "").toLowerCase() === node.type.name.toLowerCase();

    const optionRow = [
      opt.description && {
        label: "JSON Schema Type",
        elements: [{ type: "display", content: opt.description }],
      },
      opt?.enum?.length && {
        label: "Replace with",
        elements: opt.enum.map((val: string) => ({
          type: "button",
          content: val,
          onSelect: { type: "simpleSwap", payload: `"${val}"` },
        })),
      },
      ...(opt.type === "object"
        ? AnyOfObjOptionalFieldPicker(opt, node, fullCode)
        : []),
      ...(opt.type === "array" ? AnyOfArray(opt, node) : []),
      simpleType.has(opt.type) &&
        !isOfDescribedType && {
          label: "Switch to",
          elements: [
            opt.$$labeledType && {
              type: "display",
              content: opt.$$labeledType,
            },
            currentNodeType !== opt.type && {
              type: "button",
              content: `${opt.type}`,
              onSelect: {
                type: "simpleSwap",
                // payload: "null"
                payload: simpleTypeMap[opt.type],
              },
            },
          ].filter((x) => x) as MenuRow[],
        },
    ].filter((x) => x);
    return optionRow;
  });
  return rows;
};

const generateSubItem = (subItem: JSONSchema7Definition) => {
  if (typeof subItem !== "object" || subItem.type !== "object") {
    return false;
  }
  const payload = materializeAnyOfOption(subItem);
  return { content: payload, payload };
};

const ArrayItemBuilder: Component = ({ content, node }) => {
  const items = content.items;
  if (!items || typeof items === "boolean") {
    return [];
  }
  const elements = (Array.isArray(items) ? items : [items]).flatMap((item) => {
    if (typeof item === "boolean") {
      return [];
    }
    return (item.oneOf || []).flatMap((subItem) => {
      const result = generateSubItem(subItem);
      if (!result) {
        return [];
      }
      const { content, payload } = result;
      const element: MenuElement = {
        type: "button",
        content,
        onSelect: {
          type: "addElementAsSiblingInArray",
          payload,
          nodeId: nodeToId(node.parent?.lastChild?.prevSibling!),
        },
      };
      return element;
    });
  });

  return [{ label: "Insert", elements }];
};

const makeSimpleComponent: (x: string) => Component = (content) => (props) => {
  return [
    {
      label: "Inferred JSON Type",
      elements: [{ type: "display", content }],
    },
  ];
};

const GenericComponent = makeSimpleComponent("hi generic");

const PropertyNameComponent: Component = (props) => {
  const { node, fullCode } = props;
  return [
    {
      label: "PROPERTY",
      elements: [
        { type: "display", content: fullCode.slice(node.from, node.to) },
        {
          type: "button",
          content: "remove key",
          onSelect: { type: "removeObjectKey", nodeId: nodeToId(node) },
        },
        ...directionalMoves(node),
      ],
    },
  ];
};

const ObjectComponent: Component = ({ node }) => [
  {
    label: "Add Field",
    elements: [
      {
        type: "free-input",
        label: "Add field",
        onSelect: {
          type: "addObjectKey",
          nodeId: nodeToId(node),
          payload: { key: "$$INPUT_BIND", value: "null" },
        },
      },
    ],
  },
];

const ArrayComponent: Component = ({ node }) => {
  return [
    {
      label: "Add element",
      elements: [
        { label: "boolean", value: "false" },
        { label: "number", value: "0" },
        { label: "string", value: '""' },
        { label: "object", value: "{}" },
        { label: "array", value: "[]" },
      ].map(({ label, value }) => ({
        type: "button",
        content: label,
        onSelect: {
          type: "addElementAsSiblingInArray",
          payload: value,
          nodeId: nodeToId(node.lastChild?.prevSibling!),
        },
      })),
    },
  ];
};

const bracketComponent: Component = (props) =>
  ArrayComponent({ ...props, node: props.node.parent! });

const curlyBracketComponent: Component = (props) =>
  ObjectComponent({ ...props, node: props.node.parent! });

const ParentIsPropertyComponent = makeSimpleComponent("hi property");

const directionalMoves = (node: SyntaxNode): MenuElement[] => {
  const outputDirections: MenuElement[] = [];
  const bounds = boundCheck(node);
  if (!bounds.isFirst) {
    outputDirections.push({
      type: "button",
      content: "Move item left",
      onSelect: { type: "decreaseItemIdx", nodeId: nodeToId(node) },
    });
    // {
    //   type: "button",
    //   content: "Set item as Last",
    //   onSelect: { type: "moveItemToEnd" },
    // },
  }
  if (!bounds.isLast) {
    outputDirections.push({
      type: "button",
      content: "Move item right",
      onSelect: { type: "increaseItemIdx", nodeId: nodeToId(node) },
    });
    // {
    //   type: "button",
    //   content: "Set item as First",
    //   onSelect: { type: "moveItemToStart" },
    // },
  }
  return outputDirections;
};

const ParentIsArrayComponent: Component = ({ node }) => {
  return [
    {
      label: "PROPERTY",
      elements: [
        {
          type: "button",
          content: "Remove Item",

          onSelect: { type: "removeElementFromArray", nodeId: nodeToId(node) },
        },
        ...directionalMoves(node),
      ],
    },
  ];
};

const BooleanComponent: Component = ({ node }) => [
  {
    label: "Set to",
    elements: ["true", "false"].map((payload) => ({
      type: "button",
      content: payload,
      onSelect: { type: "simpleSwap", nodeId: nodeToId(node), payload },
    })),
  },
];

const menuSwitch: componentContainer = {
  EnumPicker,
  ObjPicker,
  AnyOfPicker,
  GenericComponent,
  ArrayItemBuilder,
};
const typeBasedComponents: componentContainer = {
  Object: ObjectComponent,
  PropertyName: PropertyNameComponent,
  Array: ArrayComponent,
  String: makeSimpleComponent("string"),
  Number: makeSimpleComponent("number"),
  False: BooleanComponent,
  True: BooleanComponent,
  Null: makeSimpleComponent("null"),

  ...Object.fromEntries(["[", "]"].map((el) => [el, bracketComponent])),
  ...Object.fromEntries(["{", "}"].map((el) => [el, curlyBracketComponent])),
  "⚠": makeSimpleComponent("⚠"),
};
const parentResponses: componentContainer = {
  Property: ParentIsPropertyComponent,
  Array: ParentIsArrayComponent,
};

export const liminalNodeTypes = new Set(["⚠", "{", "}", "[", "]"]);
export function retargetToAppropriateNode(
  node: SyntaxNode | SyntaxNode
): SyntaxNode | SyntaxNode {
  let targetNode = node;
  if (liminalNodeTypes.has(node.type.name)) {
    targetNode = node.parent!;
  } else if (node.type.name === "PropertyName") {
    targetNode = node.nextSibling!;
  }
  return targetNode;
}

function getSchemaForRetargetedNode(
  node: SyntaxNode,
  schemaMap: SchemaMap
): JSONSchema7 {
  let targetNode = retargetToAppropriateNode(node);

  const from = targetNode.from;
  const to = targetNode.to;

  let schemaChunk: JSONSchema7[] = schemaMap[`${from}-${to}`];
  if (schemaChunk?.length > 1) {
    return { anyOf: schemaChunk };
  } else if (schemaChunk?.length === 1) {
    return schemaChunk[0];
  }
  // implying that its not an array?
  return schemaChunk as any as JSONSchema7;
}

// todo remove first argument
export function generateMenuContent(
  currentCodeSlice: string,
  syntaxNode: SyntaxNode,
  schemaMap: SchemaMap,
  fullCode: string
): MenuRow[] {
  const schemaChunk = getSchemaForRetargetedNode(syntaxNode, schemaMap);

  //   TODO these should be functions
  const type = syntaxNode.type.name;
  let typeBasedProperty: Component | null = null;
  if (typeBasedComponents[type]) {
    typeBasedProperty = typeBasedComponents[type];
  } else if (!schemaChunk) {
    console.log("missing imp for", type);
  } else if (typeBasedProperty && !typeBasedProperty[type]) {
    console.log("missing type imp for", type);
  }

  let contentBasedItem: Component | null = null;
  // TODO work through options listed in the validate wip
  // not sure what this refers, but there are some missing options for sure, maybe explore the .type = X?
  if (schemaChunk && schemaChunk.enum) {
    contentBasedItem = menuSwitch.EnumPicker;
  } else if (schemaChunk && schemaChunk.type === "object") {
    contentBasedItem = menuSwitch.ObjPicker;
  } else if (schemaChunk && schemaChunk.anyOf) {
    contentBasedItem = menuSwitch.AnyOfPicker;
  } else if (schemaChunk && schemaChunk.type === "array") {
    contentBasedItem = menuSwitch.ArrayItemBuilder;
  }

  //   assemble the content for display
  const content: MenuRow[] = [];
  if (schemaChunk?.description) {
    content.push({
      label: "JSON Schema Type",
      elements: [{ type: "display", content: schemaChunk.description }],
    });
  }
  const componentProps: ComponentProps = {
    content: schemaChunk,
    node: syntaxNode,
    fullCode,
  };
  if (schemaChunk && !!contentBasedItem) {
    contentBasedItem(componentProps).forEach((x) => content.push(x));
  }
  if (typeBasedProperty) {
    typeBasedProperty(componentProps).forEach((x) => content.push(x));
  }

  // const parentType = syntaxNode.parent!.type.name;
  const parentType = retargetToAppropriateNode(syntaxNode).parent!.type.name;
  if (parentResponses[parentType]) {
    parentResponses[parentType](componentProps).forEach((x) => content.push(x));
  }
  const computedMenuContents = cleanSections(
    simpleMerge(content.filter((x) => x))
  );
  return computedMenuContents;
}

function deduplicate(rows: any[]): any[] {
  const hasSeen: Set<string> = new Set([]);
  return rows.filter((x) => {
    const key = JSON.stringify(x);
    if (hasSeen.has(key)) {
      return false;
    }
    hasSeen.add(key);
    return true;
  });
}

function simpleMerge(content: MenuRow[]): MenuRow[] {
  const groups = content.reduce((acc: Record<string, any[]>, row) => {
    acc[row.label] = (acc[row.label] || []).concat(row.elements);
    return acc;
  }, {});

  return Object.entries(groups).map(
    ([label, elements]) =>
      ({ label, elements: deduplicate(elements).filter((x) => x) } as MenuRow)
  );
}

function cleanSections(content: MenuRow[]): MenuRow[] {
  return content.filter((x) => x.elements.length);
}
