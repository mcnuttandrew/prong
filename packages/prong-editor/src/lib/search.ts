import { MenuRow } from "./compute-menu-contents";
import { SyntaxNode } from "@lezer/common";
// import levenshtein from 'js-levenshtein';
// todo https://itnext.io/string-similarity-the-basic-know-your-algorithms-guide-3de3d7346227

export function filterContents(searchTerm: string, rows: MenuRow[]): MenuRow[] {
  const term = searchTerm.toLowerCase();
  const result = rows
    .map((row) => {
      return {
        ...row,
        elements: row.elements.filter((el) => {
          switch (el.type) {
            case "button":
              return ((el.label || "") + el.content)
                .toLowerCase()
                .includes(term);
            case "display":
              return ((el.label || "") + el.content)
                .toLowerCase()
                .includes(term);
            case "free-input":
              return false;
            default:
              return true;
          }
        }),
      };
    })
    .filter((row) => row.elements.length > 0);
  return result;
}

/**
 * This function potentially filters the content depending on some heuristics
 * @param targetNode
 * @param fullCode
 * @param contents
 * @returns MenuRow[]
 */
export function potentiallyFilterContentForGesture(
  targetNode: SyntaxNode,
  fullCode: string,
  contents: MenuRow[]
) {
  // this suggests that this MAY be an autocomplete gesture
  const targetNodeIsError = targetNode.type.name === "âš ";
  const targNodeContent = fullCode.slice(targetNode.from, targetNode.to);
  const useContentAsFilter =
    targetNodeIsError && !targNodeContent.includes(" ");

  // console.log(
  //   targetNode.type,
  //   targNodeContent,
  //   contents
  // );

  if (useContentAsFilter) {
    return filterContents(targNodeContent, contents);
  } else {
    return contents;
  }
}
