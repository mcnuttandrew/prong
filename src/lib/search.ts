import { MenuRow } from "./compute-menu-contents";
// import levenshtein from 'js-levenshtein';
// todo https://itnext.io/string-similarity-the-basic-know-your-algorithms-guide-3de3d7346227

export function filterContents(searchTerm: string, rows: MenuRow[]): MenuRow[] {
  const result = rows
    .map((row) => {
      return {
        ...row,
        elements: row.elements.filter((el) => {
          switch (el.type) {
            case "button":
              return el.content.includes(searchTerm);
            case "display":
              return el.content.includes(searchTerm);
            case "free-input":
              return false;
            default:
              return true;
          }
        }),
      };
    })
    .filter((row) => row.elements.length > 0);
  return result;
}
