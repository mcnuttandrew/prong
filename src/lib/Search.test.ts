import { filterContents } from "./search";
import { findNodeByText } from "./test-utils";
import { generateMenuContent } from "./compute-menu-contents";
import VegaLiteV5Schema from "../constants/vega-lite-v5-schema.json";
import { createNodeMap } from "./utils";
const updatedSchema = {
  ...VegaLiteV5Schema,
  $ref: "#/definitions/Config",
};

test("filterContents", async () => {
  const exampleString = `{ "axis": { } }`;
  const schemaMap = await createNodeMap(updatedSchema, exampleString);
  const target = findNodeByText(exampleString, `{ }`)!;
  const generatedContent = generateMenuContent(
    target,
    schemaMap,
    exampleString
  );
  expect(filterContents("gr", generatedContent)).toMatchSnapshot();
});
