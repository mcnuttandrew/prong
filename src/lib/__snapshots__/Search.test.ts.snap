// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterContents 1`] = `
Array [
  Object {
    "elements": Array [
      Object {
        "content": "grid",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"grid\\"",
            "value": "true",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      Object {
        "content": "gridCap",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"gridCap\\"",
            "value": "",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      Object {
        "content": "gridColor",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"gridColor\\"",
            "value": "{ } ",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      Object {
        "content": "gridDash",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"gridDash\\"",
            "value": "[ ] ",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      Object {
        "content": "gridDashOffset",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"gridDashOffset\\"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      Object {
        "content": "gridOpacity",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"gridOpacity\\"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
      Object {
        "content": "gridWidth",
        "onSelect": Object {
          "nodeId": "10-13",
          "payload": Object {
            "key": "\\"gridWidth\\"",
            "value": "0",
          },
          "type": "addObjectKey",
        },
        "type": "button",
      },
    ],
    "label": "Add Field",
  },
]
`;
